dlt/common/metrics.py

```diff
@@ -0,0 +1,71 @@
+import datetime  # noqa: I251
+from typing import Any, Dict, List, NamedTuple, Optional, Tuple, TypedDict  # noqa: 251
+
+
+class DataWriterMetrics(NamedTuple):
+    file_path: str
+    items_count: int
+    file_size: int
+    created: float
+    last_modified: float
+
+    def __add__(self, other: Tuple[object, ...], /) -> Tuple[object, ...]:
+        if isinstance(other, DataWriterMetrics):
+            return DataWriterMetrics(
+                self.file_path if self.file_path == other.file_path else "",
+                # self.table_name if self.table_name == other.table_name else "",
+                self.items_count + other.items_count,
+                self.file_size + other.file_size,
+                min(self.created, other.created),
+                max(self.last_modified, other.last_modified),
+            )
+        return NotImplemented
+
+
+class StepMetrics(TypedDict):
+    """Metrics for particular package processed in particular pipeline step"""
+
+    started_at: datetime.datetime
+    """Start of package processing"""
+    finished_at: datetime.datetime
+    """End of package processing"""
+
+
+class ExtractDataInfo(TypedDict):
+    name: str
+    data_type: str
+
+
+class ExtractMetrics(StepMetrics):
+    schema_name: str
+    job_metrics: Dict[str, DataWriterMetrics]
+    """Metrics collected per job id during writing of job file"""
+    table_metrics: Dict[str, DataWriterMetrics]
+    """Job metrics aggregated by table"""
+    resource_metrics: Dict[str, DataWriterMetrics]
+    """Job metrics aggregated by resource"""
+    dag: List[Tuple[str, str]]
+    """A resource dag where elements of the list are graph edges"""
+    hints: Dict[str, Dict[str, Any]]
+    """Hints passed to the resources"""
+
+
+class NormalizeMetrics(StepMetrics):
+    job_metrics: Dict[str, DataWriterMetrics]
+    """Metrics collected per job id during writing of job file"""
+    table_metrics: Dict[str, DataWriterMetrics]
+    """Job metrics aggregated by table"""
+
+
+class LoadJobMetrics(NamedTuple):
```

Comment:
I think it might be of interested to also track whether the data from this job was loaded to the staging destination or the staging dataset

Reply:
It would also be cool to track if this was a table chain followup job. With these three flags you could make the/any platform UI much more legible.

Reply:
Thinking a bit more about it, maybe a job_type would be interesting too in the context of displaying data to the user. Types could be:

single table jobs
* copy (all jobs that just move files around)
* load (all jobs that move data into a database)

table chain jobs
* merge (merge jobs)
* upsert (upsert merge jobs)
* scd2 (scd2 merge)
* replace (replace via insert from staging)
* whatever we call this chunking update